### PR TITLE
Feature/label smoothing

### PR DESCRIPTION
**Pull request by blester125**
This PR adds label smoothing for pytorch seq2seq and tf2 eager mode seq2seq. AFAIK we are planning to deprecate tf1 soonish so I didn't bother adding it there.

One weird thing is that we have this `SequenceLoss` wrapper that instantiates a loss function inside of it. We can pass in different losses but if it needs a extra argument it gets weird. I created and inner function that acts like a partial application of the argument.

I also had to add `**kwargs` to all the eager loss function so that I could pass in the label smoothing for the seq2seq one. I tested all the tasks with tf eager mode and they all work

### CODE REVIEW

File: layers/eight_mile/pytorch/layers.py [hunk 1]
```diff
@@ -170,6 +170,31 @@ def extra_repr(self):
         return f"reduction={self.avg}"
 
 
+class LabelSmoothingLoss(nn.Module):
```
Comment (dpressel) [inline layers/eight_mile/pytorch/layers.py:173]: This should contain documentation and a reference

File: layers/eight_mile/tf/optz.py [hunk 2]
```diff
@@ -776,17 +776,17 @@ def weight_decay_fn():
     def global_step(self):
         return self.optimizer.iterations
 
-    def update(self, model, x, y, num_replicas=1):
+    def update(self, model, x, y, num_replicas=1, **kwargs):
```
Comment (dpressel) [inline layers/eight_mile/tf/optz.py:779]: there doesnt seem to be much API parity between TF and PyTorch with this feature.  In once case we are changing the impl of EagerOptimizer and in the other its unmodified?

File: baseline/tf/seq2seq/training/eager.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 from baseline.tf.seq2seq.training.utils import to_tensors, SHUF_BUF_SZ, NUM_PREFETCH
 
 
-def loss(model, features, labels):
+def loss(model, features, labels, **kwargs):
```
Comment (dpressel) [inline baseline/tf/seq2seq/training/eager.py:16]: I think it would be possible to produce a functor with the right underlying implementation so that this can be done prior to loss function definition (similar to the fact we do this upfront in the PyTorch impl).  Then the loss function could be defined differently depending on if its on or not.  I guess this pushes the problem up but to me it feels weird that the optimizer ends up needing kwargs everywhere to set it up right.  Would be better IMO if we could adding the kwargs to the optimizer and loss function defs

File: layers/eight_mile/tf/optz.py (hunk 2 above)
Comment (blester125) [reply layers/eight_mile/tf/optz.py:779]: There already wasn't API parity between the two. The pytorch seq2seq model is structured like all out original models (tf1 and pytorch) where there is a `.create_loss` method. TF2 Eager (the only seq2seq supported mode) is already setup differently where there is a `loss` function in the trainer that is responsible for running the model's forward pass and calculating the loss. That function is called by the eager optimizer so the only way to inject things that tweak that loss calculation is to pass them through the EagerOptimzer like this. The TF2 Eager version doesn't have something akin to `.create_loss` that we can modify or override in a subclass.

File: baseline/tf/seq2seq/training/eager.py (hunk 3 above)
Comment (blester125) [reply baseline/tf/seq2seq/training/eager.py:16]: I link the spirit of this in that it seems to be separating the loss from the model, moving towards something that we are able to specify in the config, swap out, etc but I guess I'm not in love with the loss function being responsible for running the forward part of the model, I know it is mostly there because of how different that tagger was, but I was hoping to not to a huge refactor on how eager losses are done because I still don't know if the way we are doing it is the best way

Comment (blester125) [reply baseline/tf/seq2seq/training/eager.py]: Updated to use a functor

File: baseline/tf/seq2seq/training/eager.py [hunk 4]
```diff
@@ -13,24 +13,42 @@
...
-    labels = labels[1:mx_seq_len + 1, :]
-    logits = logits[:mx_seq_len, :, :]
-    losses = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=labels)
-    loss_mask = tf.cast(tf.sequence_mask(label_lengths-1), dtype=tf.float32)
-    losses = losses * tf.transpose(loss_mask, [1, 0])
-
-    losses = tf.reduce_sum(losses)
-    losses /= tf.cast(tf.reduce_sum(label_lengths), tf.float32)
-    return losses
+class Seq2SeqLoss(tf.keras.layers.Layer):
```
Comment (dpressel) [inline baseline/tf/seq2seq/training/eager.py:16]: It looks like right now the distributed mode hasnt been updated to use this loss but it should be

Comment (dpressel) [reply baseline/tf/seq2seq/training/eager.py:16]: We can add this later. I’ll merge for now to keep it moving
